inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp

```diff
@@ -125,480 +152,753 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
+                            maxConf = conf;
+                            maxCIdx = c;
+                        }
+                    }
+                }
+            }
+            // MXNet statistic, indices and detectionCount is for each image
+            if (maxCIdx > 0) {
+                mtx.lock();
+                indicesData[off + detectionsData[n * classesNum]] = maxCIdx * priorsNum + p;  // de-refer to get prior and class id.
+                detectionsData[n * classesNum]++;
+                mtx.unlock();
+            }
+        });
+        // topk
+        // in:  indicesData, detection_count(filtered num)
+        // out: buffer, detection_count(k)
+        int count = detectionsData[n * classesNum];
+        int k = (topK == -1 ? count : (std::min)(topK, count));
+
+        const float *pconf = reorderedConfData + off;
+        int *indices = indicesData + off;
+        int *pbuffer = indicesBufData + off;
+        topk(indices, pbuffer, pconf, count, k);
+        detectionsData[n * classesNum] = k;
+    }
 }
 
-void MKLDNNDetectionOutputNode::decodeBBoxes(const float *prior_data,
-                                       const float *loc_data,
-                                       const float *variance_data,
-                                       float *decoded_bboxes,
-                                       float *decoded_bbox_sizes,
-                                       int* num_priors_actual,
+inline void MKLDNNDetectionOutputNode::decodeBBoxes(const float *priorData,
+                                       const float *locData,
+                                       const float *varianceData,
+                                       float *decodedBboxes,
+                                       float *decodedBboxSizes,
+                                       int* numPriorsActual,
                                        int n,
                                        const int& offs,
-                                       const int& pr_size,
-                                       bool decodeType) {
-    num_priors_actual[n] = _num_priors;
-    if (!_normalized && decodeType) {
-        int num = 0;
-        for (; num < _num_priors; ++num) {
-            float batch_id = prior_data[num * pr_size + 0];
-            if (batch_id == -1.f) {
-                num_priors_actual[n] = num;
-                break;
+                                       const int& priorSize,
+                                       bool decodeType,
+                                       const int *confInfoH,
+                                       const int *confInfoV) {
+    int prNum = numPriorsActual[n];
+    if (!decodeType) {
+        prNum = priorsNum;
+    }
+    parallel_for(prNum, [&](int p) {
+        if (isSparsityWorthwhile) {
+            if (isShareLoc && confInfoV[p] == -1) {
+                return;
+            }
+            if (!isShareLoc && !decreaseClassId && confInfoH[priorsNum] == 0) {
```

Comment:
This conditions seems independent from the loop iteration. Does it makes sense to extract it from the loop?

Reply:
Extracted out of parallel and act as a check if continue! thank you!